iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp

```diff
@@ -0,0 +1,339 @@
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<OpTy>(op.getLoc(), op.predicate(), args[0], args[1]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeSelectOp : public OpConversionPattern<mlir::SelectOp> {
+  using OpConversionPattern<mlir::SelectOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      mlir::SelectOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::any_of(args,
+                     [](Value arg) { return arg.getType().isIntOrFloat(); })) {
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<SelectOp>(op.getLoc(), args[0], args[1], args[2]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeYieldOp : public OpConversionPattern<linalg::YieldOp> {
+  using OpConversionPattern<linalg::YieldOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::YieldOp yieldOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!llvm::all_of(args,
+                      [](Value v) { return v.getType().isa<VectorType>(); })) {
+      return failure();
+    }
+    SmallVector<Value, 4> vecArgs;
+    for (Value arg : args) {
+      Value val = legalizeToVectorType(rewriter, arg);
+      if (!val) return failure();
+      vecArgs.push_back(val);
+    }
+    rewriter.replaceOpWithNewOp<linalg::YieldOp>(yieldOp, vecArgs);
+    return success();
+  }
+};
+
+struct VectorizeGenericOp : public OpConversionPattern<linalg::GenericOp> {
+  using OpConversionPattern<linalg::GenericOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp genericOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    // Ignore a case that either a nop pattern or a linalg.fill pattern.
+    if (isa<linalg::YieldOp>(
+            genericOp.region().front().getOperations().front())) {
+      return failure();
+    }
+
+    if (llvm::any_of(genericOp.iterator_types(), [](Attribute attr) {
+          return attr.cast<StringAttr>().getValue() !=
+                 getParallelIteratorTypeName();
+        })) {
+      return failure();
+    }
+
+    // Do not vectorize if one of the operand is 0-D or one of the operand is
+    // not iterated on contiguous memory.
+    for (auto map : genericOp.getIndexingMaps()) {
+      if (map.getNumResults() == 0) return failure();
+      int lastDimPos =
+          map.getResults().back().cast<AffineDimExpr>().getPosition();
+      for (auto expr : map.getResults()) {
+        if (expr.cast<AffineDimExpr>().getPosition() > lastDimPos) {
+          return failure();
+        }
+      }
+    }
+
+    SmallVector<IREE::PlaceholderOp, 4> operands;
+    SmallVector<MemRefType, 4> vecMemRefs;
+    for (auto operand : args) {
+      auto op = operand.getDefiningOp<IREE::PlaceholderOp>();
```

Comment:
Why not handle `iree.placeholder` ops separately in a pattern?

Reply:
There are several checks to decide if we wanna vectorize the Linalg op. If we handle it in a separately pattern, it seems that we need to verify the same thing for several times (O(n) times where n is the number of operands of Linalg op). 

I think in the end we don't need to rewrite the pattern. A better idea is to use vector.type_cast on the operands. But this would need more and more works. I think the conversion from vector to other dialects (eg SPIR-V) is not implemented yet.

Reply:
Makes sense. Thanks for explaining.